src/numerics/petsc_matrix.C

```diff
@@ -1201,7 +1201,89 @@ void PetscMatrix<T>::add (const T a_in, const SparseMatrix<T> & X_in)
 }
 
 
+template <typename T>
+void PetscMatrix<T>::matrix_matrix_mult (SparseMatrix<T> & X_in, SparseMatrix<T> & Y_out)
+{
+  libmesh_assert (this->initialized());
+
+   // sanity check.
+   libmesh_assert_equal_to (this->n(), X_in.m());
+   libmesh_assert_equal_to (this->m(), Y_out.m());
+   libmesh_assert_equal_to (X_in.n(), Y_out.n());
+
+   const PetscMatrix<T> * X = cast_ptr<const PetscMatrix<T> *> (&X_in);
+   PetscMatrix<T> * Y = cast_ptr<PetscMatrix<T> *> (&Y_out);
+
+   libmesh_assert (X);
+   libmesh_assert (Y);
+
+   PetscErrorCode ierr=0;
+
+   // the matrix from which we copy the values has to be assembled/closed
+   libmesh_assert(X->closed());
+
+   semiparallel_only();
+
+   ierr = MatMatMult(_mat, X->_mat, MAT_INITIAL_MATRIX, PETSC_DEFAULT, &Y->_mat);
+   LIBMESH_CHKERR(ierr);
+
+}
+
+template <typename T>
+void PetscMatrix<T>::add_sparse_matrix (const SparseMatrix<T> & spm,
+                        const std::map<numeric_index_type, numeric_index_type> & row_ltog,
+                        const std::map<numeric_index_type, numeric_index_type> & col_ltog,
+                        const T scalar)
+{
+  libmesh_assert_equal_to(spm.m(), row_ltog.size());
+  libmesh_assert_equal_to(spm.n(), col_ltog.size());
+  // make sure matrix has larger size than spm
+  libmesh_assert_greater_equal(this->m(), spm.m());
+  libmesh_assert_greater_equal(this->n(),spm.n());
+
+  if(!this->closed())
+    this->close();
 
+  PetscErrorCode ierr=0;
+
+  ierr=MatSetOption(_mat, MAT_NEW_NONZERO_ALLOCATION_ERR, PETSC_FALSE);
```

Comment:
@fdkong this is going to make the `add_sparse_matrix()` routine really slow in case the matrix is not pre-allocated correctly, right? Do we have no way of properly allocating the sparsity ahead of time in the application? I prefer not to allow people to do something that will make their code really slow, since then they will blame us for that :)

Reply:
Right, we should remove this line of code. In the preconditioner, we definitely can figure out the preallocation even though it is not straightforward. 

We will preallocate memory correctly before we call `add_sparse_matrix`

Reply:
Good point @jwpeterson @fdkong . I agree it would be ideal to pre-allocate memory before calling this function. I have no idea how to do that though... hopefully @fdkong can give me some directions on this.